File: built-in plugins dsp/mb-comp-multiband.cpp
/**
 * MB Multiband Comp
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : 3-band multiband compressor
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_COMP_MULTIBAND_H
#define MB_COMP_MULTIBAND_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCompMultiband : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-comp-multiband";
    static constexpr const char* PLUGIN_NAME    = "MB Multiband Comp";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float lowThresh = -20f;  // range [-60, 0]
    float midThresh = -18f;  // range [-60, 0]
    float highThresh = -16f;  // range [-60, 0]
    };

    MbCompMultiband() = default;
    ~MbCompMultiband() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.lowThresh = std::clamp(params.lowThresh, -60f, 0f);
        params.midThresh = std::clamp(params.midThresh, -60f, 0f);
        params.highThresh = std::clamp(params.highThresh, -60f, 0f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Multiband Comp
        return input;
    }
};

#endif // MB_COMP_MULTIBAND_H

File: maxbooster7.5/built-in plugins dsp/mb-comp-multiband.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbCompMultibandPlugin: PluginDefinition = {
  id: "mb-comp-multiband",
  slug: "mb-comp-multiband",
  name: "MB Multiband Comp",
  category: "effect",
  type: "compressor",
  version: "1.0.0",
  description: "3-band multiband compressor",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "lowThresh",
      name: "Low Thresh",
      type: "float",
      defaultValue: -20,
      minValue: -60,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "midThresh",
      name: "Mid Thresh",
      type: "float",
      defaultValue: -18,
      minValue: -60,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "highThresh",
      name: "High Thresh",
      type: "float",
      defaultValue: -16,
      minValue: -60,
      maxValue: 0,
      automatable: true,
    },
  ],
  defaultPreset: { lowThresh: -20, midThresh: -18, highThresh: -16 },
};

export default MbCompMultibandPlugin;
